File: fhir-bridge/node/src/evageneClient.ts
/**
 * Thin client for the subset of the Evagene REST API the bridge uses.
 *
 * One method per endpoint. Orchestration lives in the callers; this
 * class only translates method arguments into HTTP calls and HTTP
 * responses into typed return values.
 */

import type { HttpGateway, HttpMethod } from './httpGateway.js';
import type { BiologicalSex, RelativeType } from './intakeFamily.js';

const HTTP_OK_LOWER = 200;
const HTTP_OK_UPPER = 300;

export class EvageneApiError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'EvageneApiError';
  }
}

export interface EvageneClientOptions {
  readonly baseUrl: string;
  readonly apiKey: string;
  readonly http: HttpGateway;
}

export interface CreatePedigreeArgs {
  readonly displayName: string;
}

export interface CreateIndividualArgs {
  readonly displayName: string;
  readonly biologicalSex: BiologicalSex;
  readonly yearOfBirth?: number;
}

export interface AddRelativeArgs {
  readonly pedigreeId: string;
  readonly relativeOf: string;
  readonly relativeType: RelativeType;
  readonly displayName: string;
  readonly biologicalSex: BiologicalSex;
  readonly yearOfBirth?: number;
}

export interface EvageneApi {
  getPedigreeDetail(pedigreeId: string): Promise<Record<string, unknown>>;
  createPedigree(args: CreatePedigreeArgs): Promise<string>;
  createIndividual(args: CreateIndividualArgs): Promise<string>;
  addIndividualToPedigree(pedigreeId: string, individualId: string): Promise<void>;
  designateAsProband(individualId: string): Promise<void>;
  addRelative(args: AddRelativeArgs): Promise<string>;
  deletePedigree(pedigreeId: string): Promise<void>;
}

export class EvageneClient implements EvageneApi {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly http: HttpGateway;

  constructor(options: EvageneClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/$/, '');
    this.apiKey = options.apiKey;
    this.http = options.http;
  }

  async getPedigreeDetail(pedigreeId: string): Promise<Record<string, unknown>> {
    return this.request('GET', `/api/pedigrees/${pedigreeId}`);
  }

  async createPedigree(args: CreatePedigreeArgs): Promise<string> {
    const payload = await this.request('POST', '/api/pedigrees', {
      display_name: args.displayName,
    });
    return requireStringField(payload, 'id');
  }

  async createIndividual(args: CreateIndividualArgs): Promise<string> {
    const body: Record<string, unknown> = {
      display_name: args.displayName,
      biological_sex: args.biologicalSex,
    };
    withYearOfBirth(body, args.yearOfBirth);
    const payload = await this.request('POST', '/api/individuals', body);
    return requireStringField(payload, 'id');
  }

  async addIndividualToPedigree(pedigreeId: string, individualId: string): Promise<void> {
    await this.sendIgnoringBody(
      'POST',
      `/api/pedigrees/${pedigreeId}/individuals/${individualId}`,
      {},
    );
  }

  async designateAsProband(individualId: string): Promise<void> {
    await this.sendIgnoringBody('PATCH', `/api/individuals/${individualId}`, { proband: 1 });
  }

  async addRelative(args: AddRelativeArgs): Promise<string> {
    const body: Record<string, unknown> = {
      relative_of: args.relativeOf,
      relative_type: args.relativeType,
      display_name: args.displayName,
      biological_sex: args.biologicalSex,
    };
    withYearOfBirth(body, args.yearOfBirth);
    const payload = await this.request(
      'POST',
      `/api/pedigrees/${args.pedigreeId}/register/add-relative`,
      body,
    );
    const individual = requireObjectField(payload, 'individual');
    return requireStringField(individual, 'id');
  }

  async deletePedigree(pedigreeId: string): Promise<void> {
    await this.sendIgnoringBody('DELETE', `/api/pedigrees/${pedigreeId}`);
  }

  private async request(
    method: HttpMethod,
    path: string,
    body?: unknown,
  ): Promise<Record<string, unknown>> {
    const response = await this.send(method, path, body);
    return parseObjectBody(response, `${method} ${path}`);
  }

  private async sendIgnoringBody(
    method: HttpMethod,
    path: string,
    body?: unknown,
  ): Promise<void> {
    await this.send(method, path, body);
  }

  private async send(method: HttpMethod, path: string, body: unknown) {
    const url = `${this.baseUrl}${path}`;
    const response = await this.http.send(
      body === undefined
        ? { method, url, headers: this.headers() }
        : { method, url, headers: this.headers(), body },
    );
    if (response.status < HTTP_OK_LOWER || response.status >= HTTP_OK_UPPER) {
      throw new EvageneApiError(
        `Evagene API returned HTTP ${response.status.toString()} for ${method} ${path}`,
      );
    }
    return response;
  }

  private headers(): Record<string, string> {
    return {
      'X-API-Key': this.apiKey,
      'Content-Type': 'application/json',
      Accept: 'application/json',
    };
  }
}

async function parseObjectBody(
  response: { json(): Promise<unknown> },
  label: string,
): Promise<Record<string, unknown>> {
  let payload: unknown;
  try {
    payload = await response.json();
  } catch (error) {
    const reason = error instanceof Error ? error.message : String(error);
    throw new EvageneApiError(`Evagene API returned non-JSON body for ${label}: ${reason}`);
  }
  if (typeof payload !== 'object' || payload === null || Array.isArray(payload)) {
    throw new EvageneApiError(`Evagene API returned non-object JSON for ${label}`);
  }
  return payload as Record<string, unknown>;
}

function withYearOfBirth(body: Record<string, unknown>, year: number | undefined): void {
  if (year === undefined) {
    return;
  }
  body.properties = { year_of_birth: year };
}

function requireStringField(payload: Record<string, unknown>, key: string): string {
  const value = payload[key];
  if (typeof value !== 'string') {
    throw new EvageneApiError(`Evagene response is missing string field '${key}'`);
  }
  return value;
}

function requireObjectField(
  payload: Record<string, unknown>,
  key: string,
): Record<string, unknown> {
  const value = payload[key];
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new EvageneApiError(`Evagene response is missing object field '${key}'`);
  }
  return value as Record<string, unknown>;
}

File: archive-triage-runner/python/src/archive_triage/evagene_client.py
"""Thin client for the subset of the Evagene REST API used by this demo.

One method per endpoint; no orchestration (that lives in
:class:`TriageService`).  Depends on an :class:`HttpGateway` so tests can
supply a fake.
"""

from __future__ import annotations

from typing import Any, Protocol

from .http_gateway import HttpGateway, HttpMethod

_HTTP_OK_LOWER = 200
_HTTP_OK_UPPER = 300


class EvageneApiError(RuntimeError):
    """Raised when the Evagene API is unreachable or returns a non-2xx response."""


class EvageneApi(Protocol):
    """The surface :class:`TriageService` depends on."""

    def create_pedigree(self, display_name: str) -> str: ...

    def import_gedcom(self, pedigree_id: str, gedcom_text: str) -> None: ...

    def has_proband(self, pedigree_id: str) -> bool: ...

    def calculate_nice(self, pedigree_id: str) -> dict[str, Any]: ...

    def delete_pedigree(self, pedigree_id: str) -> None: ...


class EvageneClient:
    def __init__(self, *, base_url: str, api_key: str, http: HttpGateway) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._http = http

    def create_pedigree(self, display_name: str) -> str:
        payload = self._request_object("POST", "/api/pedigrees", {"display_name": display_name})
        return _require_str(payload, "id")

    def import_gedcom(self, pedigree_id: str, gedcom_text: str) -> None:
        # NOTE: Evagene's GEDCOM import takes a JSON body with a "content"
        # field, not raw text/plain — see https://evagene.net/docs.
        self._request(
            "POST",
            f"/api/pedigrees/{pedigree_id}/import/gedcom",
            body={"content": gedcom_text},
        )

    def has_proband(self, pedigree_id: str) -> bool:
        payload = self._request_object("GET", f"/api/pedigrees/{pedigree_id}", body=None)
        individuals = payload.get("individuals")
        if not isinstance(individuals, list):
            return False
        return any(_is_proband(member) for member in individuals)

    def calculate_nice(self, pedigree_id: str) -> dict[str, Any]:
        return self._request_object(
            "POST",
            f"/api/pedigrees/{pedigree_id}/risk/calculate",
            body={"model": "NICE"},
        )

    def delete_pedigree(self, pedigree_id: str) -> None:
        self._request("DELETE", f"/api/pedigrees/{pedigree_id}", body=None)

    def _request_object(
        self,
        method: HttpMethod,
        path: str,
        body: dict[str, Any] | None,
    ) -> dict[str, Any]:
        response = self._request(method, path, body=body)
        payload = response.json()
        if not isinstance(payload, dict):
            kind = type(payload).__name__
            raise EvageneApiError(
                f"Evagene API returned non-object JSON for {method} {path}: {kind}"
            )
        return payload

    def _request(
        self,
        method: HttpMethod,
        path: str,
        *,
        body: dict[str, Any] | None,
    ) -> Any:
        url = f"{self._base_url}{path}"
        try:
            response = self._http.send(method, url, headers=self._headers(), body=body)
        except OSError as error:
            raise EvageneApiError(
                f"Evagene API unreachable for {method} {path}: {error}"
            ) from error
        if not _HTTP_OK_LOWER <= response.status_code < _HTTP_OK_UPPER:
            raise EvageneApiError(
                f"Evagene API returned HTTP {response.status_code} for {method} {path}"
            )
        return response

    def _headers(self) -> dict[str, str]:
        return {
            "X-API-Key": self._api_key,
            "Content-Type": "application/json",
            "Accept": "application/json",
        }


def _is_proband(member: Any) -> bool:
    if not isinstance(member, dict):
        return False
    value = member.get("proband")
    return isinstance(value, int | float) and value > 0


def _require_str(payload: dict[str, Any], key: str) -> str:
    value = payload.get(key)
    if not isinstance(value, str):
        raise EvageneApiError(f"Evagene response is missing string field {key!r}")
    return value
